File: templates/skills/workflow/scripts/planning_gates.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
planning-side gates helper。

沉淀 discussion / UX / spec review 的确定性触发规则与 artifact 校验。
"""

from __future__ import annotations

import json
import os
import re
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional

UI_KEYWORDS_REGEX = re.compile(
    r"页面|界面|表单|列表|面板|弹窗|导航|路由|仪表盘|编辑器|sidebar|tab|modal|dashboard|GUI|桌面|desktop|窗口|window",
    re.IGNORECASE,
)
UI_BROAD_KEYWORDS_REGEX = re.compile(
    r"UI|界面|页面|组件|布局|样式|交互|显示|渲染|视图|前端",
    re.IGNORECASE,
)
WORKSPACE_KEYWORDS_REGEX = re.compile(r"同步|sync|agent|workspace|工作区|目录", re.IGNORECASE)


def should_run_discussion(
    requirement_content: str,
    requirement_source: str,
    no_discuss: bool = False,
    gap_count: int = 0,
) -> bool:
    if no_discuss:
        return False
    if requirement_source == "inline" and len((requirement_content or "").strip()) <= 100 and gap_count == 0:
        return False
    return True


def estimate_gap_count(requirement_content: str, requirement_source: str) -> int:
    content = (requirement_content or "").strip()
    if requirement_source == "inline" and len(content) <= 100:
        return 0
    return 1 if content else 0


def build_discussion_artifact(
    requirement_source: str,
    clarifications: Optional[List[Dict[str, Any]]] = None,
    selected_approach: Optional[Dict[str, Any]] = None,
    unresolved_dependencies: Optional[List[Dict[str, Any]]] = None,
) -> Dict[str, Any]:
    return {
        "requirementSource": requirement_source,
        "clarifications": clarifications or [],
        "selectedApproach": selected_approach,
        "unresolvedDependencies": unresolved_dependencies or [],
    }


def should_run_ux_design_gate(
    requirement_content: str,
    analysis_patterns: Optional[List[Dict[str, Any]]] = None,
    discussion_artifact: Optional[Dict[str, Any]] = None,
) -> bool:
    content = requirement_content or ""
    if UI_KEYWORDS_REGEX.search(content):
        return True

    patterns = analysis_patterns or []
    has_frontend = any(
        re.search(r"react|vue|angular|svelte|tauri|electron|next\.?js|nuxt|vite", str((pattern or {}).get("name", "")), re.IGNORECASE)
        for pattern in patterns
    )
    if has_frontend and UI_BROAD_KEYWORDS_REGEX.search(content):
        return True

    clarifications = (discussion_artifact or {}).get("clarifications") or []
    return any(
        clarification.get("dimension") in {"behavior", "edge-case"}
        for clarification in clarifications
    )


def needs_workspace_detection(requirement_content: str) -> bool:
    return bool(WORKSPACE_KEYWORDS_REGEX.search(requirement_content or ""))


def detect_agent_workspaces(home_dir: Optional[str] = None) -> List[Dict[str, Any]]:
    home = Path(home_dir or Path.home())
    cursor_candidates = [
        home / ".cursor",
        home / ".config" / "Cursor",
        home / "AppData" / "Roaming" / "Cursor",
    ]
    cursor_path = next((candidate for candidate in cursor_candidates if candidate.exists()), cursor_candidates[0])
    return [
        {
            "agent": "claude-code",
            "path": str(home / ".claude"),
            "detected": (home / ".claude").exists(),
        },
        {
            "agent": "cursor",
            "path": str(cursor_path),
            "detected": cursor_path.exists(),
        },
        {
            "agent": "codex",
            "path": str(home / ".codex"),
            "detected": (home / ".codex").exists(),
        },
    ]


def validate_ux_artifact(artifact: Dict[str, Any]) -> Dict[str, Any]:
    flowchart = artifact.get("flowchart") or {}
    scenarios = flowchart.get("scenarios") or []
    pages = ((artifact.get("pageHierarchy") or {}).get("pages") or [])
    missing = []

    if len(scenarios) < 3:
        missing.append("flowchart_scenarios")

    l0_count = sum(1 for page in pages if page.get("level") == "L0")
    if l0_count > 4:
        missing.append("l0_overflow")

    return {
        "ok": not missing,
        "missing": missing,
        "scenario_count": len(scenarios),
        "page_count": len(pages),
    }


def build_spec_review_summary(spec_content: str) -> str:
    sections: List[str] = []
    for heading in ("## 2. Scope", "## 3. Constraints", "## 7. Acceptance Criteria"):
        pattern = re.compile(rf"^{re.escape(heading)}\s*$([\s\S]*?)(?=^##\s+|\Z)", re.MULTILINE)
        match = pattern.search(spec_content or "")
        if match:
            sections.append(match.group(0).strip())
    return "\n\n".join(sections)


def map_spec_review_choice(choice: str) -> Dict[str, Any]:
    mapping = {
        "Spec 正确，继续": {
            "status": "approved",
            "next_action": "continue_to_plan_generation",
            "workflow_status": "planning",
        },
        "需要修改 Spec": {
            "status": "revise_required",
            "next_action": "return_to_phase_1_spec_generation",
            "workflow_status": "spec_review",
        },
        "页面分层需要调整": {
            "status": "revise_required",
            "next_action": "return_to_phase_0_3_ux_design_gate",
            "workflow_status": "spec_review",
        },
        "缺少用户流程": {
            "status": "revise_required",
            "next_action": "return_to_phase_0_3_ux_design_gate",
            "workflow_status": "spec_review",
        },
        "需要拆分范围": {
            "status": "rejected",
            "next_action": "split_scope",
            "workflow_status": "spec_review",
        },
    }
    return mapping.get(choice, {
        "status": "pending",
        "next_action": None,
        "workflow_status": "spec_review",
    })


def main() -> int:
    import argparse

    parser = argparse.ArgumentParser(description="planning-side gates helper")
    sub = parser.add_subparsers(dest="command")

    p_discuss = sub.add_parser("discussion", help="判断是否需要 discussion")
    p_discuss.add_argument("content")
    p_discuss.add_argument("--source", default="inline")
    p_discuss.add_argument("--no-discuss", action="store_true")
    p_discuss.add_argument("--gap-count", type=int, default=0)

    p_ux = sub.add_parser("ux-gate", help="判断是否需要 UX gate")
    p_ux.add_argument("content")
    p_ux.add_argument("--patterns-json", default="[]")
    p_ux.add_argument("--discussion-json", default="{}")

    p_workspaces = sub.add_parser("workspaces", help="探测 agent 工作目录")
    p_workspaces.add_argument("--home")

    p_review = sub.add_parser("spec-review-choice", help="映射 spec review 选择")
    p_review.add_argument("choice")

    args = parser.parse_args()

    if args.command == "discussion":
        print(json.dumps({
            "run": should_run_discussion(
                args.content,
                args.source,
                no_discuss=args.no_discuss,
                gap_count=args.gap_count,
            )
        }, ensure_ascii=False))
        return 0

    if args.command == "ux-gate":
        patterns = json.loads(args.patterns_json)
        discussion = json.loads(args.discussion_json)
        print(json.dumps({
            "run": should_run_ux_design_gate(args.content, patterns, discussion)
        }, ensure_ascii=False))
        return 0

    if args.command == "workspaces":
        print(json.dumps({"workspaces": detect_agent_workspaces(args.home)}, ensure_ascii=False))
        return 0

    if args.command == "spec-review-choice":
        print(json.dumps(map_spec_review_choice(args.choice), ensure_ascii=False))
        return 0

    parser.print_help()
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: core/utils/workflow/planning_gates.js
#!/usr/bin/env node
/** @file 规划阶段门控逻辑 - Spec 审查映射、角色信号推导、Codex 审查触发、工作区检测 */

const fs = require('fs')
const os = require('os')
const path = require('path')
const { classifyRoleSignals } = require('./role_injection')

/**
 * 从需求内容中推导角色信号（委托给 role_injection 模块）
 * @param {string} requirementContent - 需求内容文本
 * @param {Array} [analysisPatterns=[]] - 项目分析模式
 * @param {Object|null} [discussionArtifact=null] - 讨论阶段产物
 * @param {Object} [extra={}] - 额外上下文信息
 * @returns {Object} 角色信号分类结果
 */
function deriveRoleSignals(requirementContent, analysisPatterns = [], discussionArtifact = null, extra = {}) {
  return classifyRoleSignals(requirementContent, analysisPatterns, discussionArtifact, extra)
}

/**
 * 检测系统中已安装的 AI 编码工具工作区
 * @param {string} homeDir - 用户主目录路径
 * @returns {Array<Object>} 检测到的工作区列表，每项包含 agent、path、detected
 */
function detectAgentWorkspaces(homeDir) {
  const home = path.resolve(homeDir || os.homedir())
  const cursorCandidates = [path.join(home, '.cursor'), path.join(home, '.config', 'Cursor'), path.join(home, 'AppData', 'Roaming', 'Cursor')]
  const cursorPath = cursorCandidates.find((candidate) => fs.existsSync(candidate)) || cursorCandidates[0]
  return [
    { agent: 'claude-code', path: path.join(home, '.claude'), detected: fs.existsSync(path.join(home, '.claude')) },
    { agent: 'cursor', path: cursorPath, detected: fs.existsSync(cursorPath) },
    { agent: 'codex', path: path.join(home, '.codex'), detected: fs.existsSync(path.join(home, '.codex')) },
  ]
}

/**
 * 将用户的 Spec 审查选择映射为状态和下一步动作
 * approve → workflow_status 'planned'（与 cmdPlan/cmdSpecReview 实际推进一致——两个消费者在
 * approve 分支用 status==='approved' 判定并推 planned，本字段仅 non-approve 分支被读取；
 * 保持字段自洽，防未来消费者直读 workflow_status 落错状态）。
 * @param {string} choice - 用户选择的审查结论文本
 * @returns {Object} 包含 status、next_action、workflow_status 的映射结果
 */
function mapSpecReviewChoice(choice) {
  return {
    'Spec 正确，生成 Plan': { status: 'approved', next_action: 'continue_to_plan_generation', workflow_status: 'planned' },
    'Spec 正确，继续': { status: 'approved', next_action: 'continue_to_plan_generation', workflow_status: 'planned' },
    '需要修改 Spec': { status: 'revise_required', next_action: 'return_to_phase_1_spec_generation', workflow_status: 'spec_review' },
    '页面分层需要调整': { status: 'revise_required', next_action: 'return_to_phase_1_spec_generation', workflow_status: 'spec_review' },
    '缺少用户流程': { status: 'revise_required', next_action: 'return_to_phase_1_spec_generation', workflow_status: 'spec_review' },
    '缺少需求细节': { status: 'revise_required', next_action: 'return_to_phase_1_spec_generation_preserve_requirement_details', workflow_status: 'spec_review' },
    '需要拆分范围': { status: 'rejected', next_action: 'split_scope', workflow_status: 'idle' },
  }[choice] || { status: 'pending', next_action: null, workflow_status: 'spec_review' }
}

/**
 * FR-6：机器 review（codex spec/plan + execute 两段 review）默认不自动触发，
 * 仅在 project-config.json 显式开启或命令 flag 显式开启时才走原 signal 判定。
 * 沿用现有 `config.workflow.<flag>` 读取范式（参见 triage_rules workflow.triageDenylist /
 * readiness workflow.readiness）。
 *
 * config 形态（任一为真即开启 codex 自动 review）：
 *   { workflow: { review: { codex: true } } }   // 细粒度
 *   { workflow: { review: true } }                // 整体开关
 *
 * @param {Object|null} config - project-config.json 内容
 * @returns {boolean}
 */
function isMachineReviewEnabled(config) {
  const review = config && config.workflow && config.workflow.review
  if (review === true) return true
  if (review && typeof review === 'object') return review.codex === true || review.enabled === true
  return false
}

/**
 * 判断是否需要执行 Codex Spec 审查（advisory-to-human）
 *
 * FR-6 降级：默认 return false（review 自动触发已改为显式开关）；
 * 仅当 `options.reviewEnabled === true`（config/flag 显式开启）时才走原 signal 判定。
 * 注意：降级的是「自动触发」，不是删除 review 能力——显式开启后两段判定完整恢复。
 *
 * @param {string} specContent - Spec 文档内容
 * @param {Object} [signals={}] - deriveRoleSignals 输出的结构化信号
 * @param {Object} [options={}] - { reviewEnabled?: boolean } 显式开关
 * @returns {{ run: boolean, reason: string|null }} 是否触发及触发原因
 */
function shouldRunCodexSpecReview(specContent, signals = {}, options = {}) {
  if (options.reviewEnabled !== true) return { run: false, reason: 'review-gating-disabled' }
  if (signals.security) return { run: true, reason: 'signal:security' }
  if (signals.backend_heavy) return { run: true, reason: 'signal:backend_heavy' }
  if (signals.data) return { run: true, reason: 'signal:data' }
  const SUPPLEMENT_REGEX = /migration|transaction|concurrency|rate.?limit/i
  if (SUPPLEMENT_REGEX.test(String(specContent || ''))) {
    const match = String(specContent || '').match(SUPPLEMENT_REGEX)
    return { run: true, reason: `regex:${match[0].toLowerCase()}` }
  }
  return { run: false, reason: null }
}

const UI_KEYWORDS_REGEX = /页面|界面|表单|列表|面板|弹窗|导航|路由|仪表盘|编辑器|sidebar|tab|modal|dashboard|GUI|桌面|desktop|窗口|window/i
const UI_BROAD_KEYWORDS_REGEX = /UI|界面|页面|组件|布局|样式|交互|显示|渲染|视图|前端/i
const FRONTEND_FRAMEWORK_REGEX = /react|vue|angular|svelte|tauri|electron|next\.?js|nuxt|vite/i
const UX_CLARIFICATION_DIMENSIONS = new Set(['behavior', 'ui', 'ux', 'interaction', 'user-facing'])

/**
 * 判断是否需要进入 discussion（需求澄清）阶段
 * @param {string} requirementText - 需求文本
 * @param {string} requirementSource - 需求来源（'inline' / 文件路径）
 * @param {boolean} noDiscuss - 用户是否显式跳过 discussion
 * @param {number} gapCount - 检测到的需求缺口计数
 * @returns {boolean}
 */
function shouldRunDiscussion(requirementText, requirementSource, noDiscuss, gapCount) {
  if (noDiscuss) return false
  const trimmed = String(requirementText || '').trim()
  const shortInline = requirementSource === 'inline' && trimmed.length <= 100 && (gapCount || 0) === 0
  return !shortInline
}

/**
 * 判断是否需要触发 UX design gate
 * @param {string} requirementText - 需求文本
 * @param {Array<Object>} analysisPatterns - 项目分析模式（含 frontend framework 的 { name } 列表）
 * @param {Object|null} discussionArtifact - 讨论阶段产物，含 clarifications[]
 * @returns {boolean}
 */
function shouldRunUxDesignGate(requirementText, analysisPatterns = [], discussionArtifact = null) {
  const text = String(requirementText || '')
  if (UI_KEYWORDS_REGEX.test(text)) return true
  const hasFrontend = (analysisPatterns || []).some((p) => FRONTEND_FRAMEWORK_REGEX.test(String((p || {}).name || '')))
  if (hasFrontend && UI_BROAD_KEYWORDS_REGEX.test(text)) return true
  if (hasFrontend && discussionArtifact) {
    const clarifications = Array.isArray(discussionArtifact.clarifications) ? discussionArtifact.clarifications : []
    if (clarifications.some((c) => c && UX_CLARIFICATION_DIMENSIONS.has(String(c.dimension || '').toLowerCase()))) return true
  }
  return false
}

/**
 * 判断是否需要执行 Codex Plan 审查（bounded-autofix）
 *
 * FR-6 降级：默认 return false；仅当 `options.reviewEnabled === true` 时走原 signal 判定。
 *
 * @param {string} planContent
 * @param {string} specContent
 * @param {Object} [signals={}]
 * @param {Object} [options={}] - { reviewEnabled?: boolean } 显式开关
 */
function shouldRunCodexPlanReview(planContent, specContent, signals = {}, options = {}) {
  if (options.reviewEnabled !== true) return { run: false, reason: 'review-gating-disabled' }
  if (signals.security) return { run: true, reason: 'signal:security' }
  if (signals.backend_heavy) return { run: true, reason: 'signal:backend_heavy' }
  if (signals.data) return { run: true, reason: 'signal:data' }
  const SUPPLEMENT_REGEX = /migration|transaction|rollback|queue|worker|cron|webhook|oauth|jwt|rbac/i
  if (SUPPLEMENT_REGEX.test(String(planContent || ''))) {
    const match = String(planContent || '').match(SUPPLEMENT_REGEX)
    return { run: true, reason: `regex:${match[0].toLowerCase()}` }
  }
  return { run: false, reason: null }
}

function main() {
  const args = [...process.argv.slice(2)]
  const command = args.shift()
  if (command === 'workspaces') {
    const homeIndex = args.indexOf('--home')
    process.stdout.write(`${JSON.stringify({ workspaces: detectAgentWorkspaces(homeIndex >= 0 ? args[homeIndex + 1] : undefined) })}\n`)
    return
  }
  if (command === 'spec-review-choice') {
    process.stdout.write(`${JSON.stringify(mapSpecReviewChoice(args.shift()))}\n`)
    return
  }
  if (command === 'codex-spec-review') {
    const content = args.shift() || ''
    const signalsIndex = args.indexOf('--signals-json')
    const signals = signalsIndex >= 0 ? JSON.parse(args[signalsIndex + 1]) : {}
    const reviewEnabled = args.includes('--review-enabled')
    process.stdout.write(`${JSON.stringify(shouldRunCodexSpecReview(content, signals, { reviewEnabled }))}\n`)
    return
  }
  if (command === 'codex-plan-review') {
    const planContent = args.shift() || ''
    const specContent = args.shift() || ''
    const signalsIndex = args.indexOf('--signals-json')
    const signals = signalsIndex >= 0 ? JSON.parse(args[signalsIndex + 1]) : {}
    const reviewEnabled = args.includes('--review-enabled')
    process.stdout.write(`${JSON.stringify(shouldRunCodexPlanReview(planContent, specContent, signals, { reviewEnabled }))}\n`)
    return
  }
  process.stderr.write('Usage: node planning_gates.js <workspaces|spec-review-choice|codex-spec-review|codex-plan-review> ...\n')
  process.exitCode = 1
}

module.exports = {
  deriveRoleSignals,
  detectAgentWorkspaces,
  mapSpecReviewChoice,
  isMachineReviewEnabled,
  shouldRunCodexSpecReview,
  shouldRunCodexPlanReview,
  shouldRunDiscussion,
  shouldRunUxDesignGate,
  UI_KEYWORDS_REGEX,
  UI_BROAD_KEYWORDS_REGEX,
  FRONTEND_FRAMEWORK_REGEX,
}

if (require.main === module) main()
